src/ray/raylet/scheduling/cluster_resource_data.cc

```diff
@@ -339,311 +93,49 @@ bool NodeResources::operator!=(const NodeResources &other) const {
-    if (predefined_resources[i] != other.predefined_resources[i]) {
-      return false;
-    }
-  }
-  for (const auto &entry : custom_resources) {
-    auto iter = other.custom_resources.find(entry.first);
-    if (iter == other.custom_resources.end() || entry.second != iter->second) {
-      return false;
-    }
-  }
-  return true;
-}
-
-bool ResourceRequest::operator!=(const ResourceRequest &other) const {
-  return !(*this == other);
-}
-
-std::string ResourceRequest::DebugString() const {
-  std::stringstream buffer;
-  buffer << " {";
-  for (size_t i = 0; i < this->predefined_resources.size(); i++) {
-    buffer << "(" << this->predefined_resources[i] << ") ";
-  }
-  buffer << "}";
-
-  buffer << "  [";
-  for (auto &it : this->custom_resources) {
-    buffer << it.first << ":"
-           << "(" << it.second << ") ";
-  }
-  buffer << "]" << std::endl;
-  return buffer.str();
-}
-
-bool TaskResourceInstances::IsEmpty() const {
-  // Check whether all resource instances of a task are zero.
-  for (const auto &predefined_resource : predefined_resources) {
-    for (const auto &predefined_resource_instance : predefined_resource) {
-      if (predefined_resource_instance != 0) {
-        return false;
-      }
-    }
-  }
-
-  for (const auto &custom_resource : custom_resources) {
-    for (const auto &custom_resource_instances : custom_resource.second) {
-      if (custom_resource_instances != 0) {
-        return false;
-      }
-    }
-  }
-  return true;
-}
-
-std::string TaskResourceInstances::DebugString() const {
-  std::stringstream buffer;
-  buffer << std::endl << "  Allocation: {";
-  for (size_t i = 0; i < this->predefined_resources.size(); i++) {
-    buffer << VectorToString(this->predefined_resources[i]);
-  }
-  buffer << "}";
-
-  buffer << "  [";
-  for (auto it = this->custom_resources.begin(); it != this->custom_resources.end();
-       ++it) {
-    buffer << ResourceID(it->first).Binary() << ":" << VectorToString(it->second) << ", ";
-  }
-
-  buffer << "]" << std::endl;
-  return buffer.str();
-}
-
-bool EqualVectors(const std::vector<FixedPoint> &v1, const std::vector<FixedPoint> &v2) {
-  return (v1.size() == v2.size() && std::equal(v1.begin(), v1.end(), v2.begin()));
-}
-
-bool TaskResourceInstances::operator==(const TaskResourceInstances &other) {
-  for (size_t i = 0; i < PredefinedResources_MAX; i++) {
-    if (!EqualVectors(this->predefined_resources[i], other.predefined_resources[i])) {
-      return false;
-    }
-  }
-
-  if (this->custom_resources.size() != other.custom_resources.size()) {
-    return false;
-  }
-
-  for (auto it1 = this->custom_resources.begin(); it1 != this->custom_resources.end();
-       ++it1) {
-    auto it2 = other.custom_resources.find(it1->first);
-    if (it2 == other.custom_resources.end()) {
-      return false;
-    }
-    if (!EqualVectors(it1->second, it2->second)) {
-      return false;
-    }
-  }
-  return true;
+  return total.Has(id);
```

Comment:
Maybe we could rename to `Contains` to be consistent with the codes elsewhere.

Reply:
Searched a bit. "Contains" doesn't seem to be widely used either. 
Considering "Has" is also pretty clear. I'm in favor of keeping "Has".